src/smt/optimization_solver.cpp

```diff
@@ -0,0 +1,137 @@
+/*********************                                                        */
+/*! \file optimization_solver.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Michael Chang
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The solver for optimization queries
+ **/
+
+#include "smt/optimization_solver.h"
+
+#include "options/smt_options.h"
+#include "smt/smt_engine.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+#include "theory/smt_engine_subsolver.h"
+
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace smt {
+
+/**
+ * d_activatedObjective is initialized to a default objective:
+ * default objective constructed with Null Node and OBJECTIVE_UNDEFINED
+ *
+ * d_savedValue is initialized to a default node (Null Node)
+ */
+
+OptimizationSolver::OptimizationSolver(SmtEngine* parent)
+    : d_parent(parent),
+      d_activatedObjective(Node(), OBJECTIVE_UNDEFINED),
+      d_savedValue()
+{
+}
+
+OptimizationSolver::~OptimizationSolver() {}
+
+OptResult OptimizationSolver::checkOpt()
+{
+  // Make sure that the objective is not the default one
+  Assert(d_activatedObjective.getType() != OBJECTIVE_UNDEFINED);
+  Assert(!d_activatedObjective.getNode().isNull());
+
+  // the smt engine to which we send intermediate queries
+  // for the linear search.
+  std::unique_ptr<SmtEngine> optChecker;
+  initializeSubsolver(optChecker);
+  NodeManager* nm = optChecker->getNodeManager();
+
+  // we need to be in incremental mode for multiple objectives since we need to
+  // push pop we need to produce models to inrement on our objective
+  optChecker->setOption("incremental", "true");
+  optChecker->setOption("produce-models", "true");
+
+  // Move assertions from the parent solver to the subsolver
+  std::vector<Node> axioms = d_parent->getExpandedAssertions();
```

Comment:
we shouldn't be calling them axioms.
Give a better name, like `parent_assertions` or `p_assertions`